driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -483,24 +488,62 @@ void setKeyspace(String keyspace) throws ConnectionException {
         }
     }
 
-    ListenableFuture<Void> setKeyspaceAsync(final String keyspace) throws ConnectionException, BusyConnectionException {
-        logger.trace("{} Setting keyspace {}", this, keyspace);
-        // Note: we quote the keyspace below, because the name is the one coming from Cassandra, so it's in the right case already
-        Future future = write(new Requests.Query("USE \"" + keyspace + '"'));
-        return Futures.transform(future, new AsyncFunction<Message.Response, Void>() {
-            @Override
-            public ListenableFuture<Void> apply(Message.Response response) throws Exception {
-                if (response instanceof SetKeyspace) {
-                    Connection.this.keyspace = ((SetKeyspace) response).keyspace;
-                    return MoreFutures.VOID_SUCCESS;
-                } else if (response.type == ERROR) {
-                    Responses.Error error = (Responses.Error) response;
-                    throw defunct(error.asException(address));
-                } else {
-                    throw defunct(new DriverInternalError("Unexpected response while setting keyspace: " + response));
+    ListenableFuture<Connection> setKeyspaceAsync(final String keyspace) throws ConnectionException, BusyConnectionException {
+        if (Objects.equal(keyspace(), keyspace)) {
+            return thisFuture;
+        }
+
+        final SettableFuture<Connection> ksFuture = SettableFuture.create();
+        final SetKeyspaceAttempt attempt = new SetKeyspaceAttempt(keyspace, ksFuture);
+
+        // Check for an existing keyspace attempt.
+        while (true) {
+            SetKeyspaceAttempt existingAttempt = targetKeyspace.get();
+            // if existing attempts' keyspace matches what we are trying to set, use it.
+            if (attempt.equals(existingAttempt)) {
+                return existingAttempt.future;
+            } else if (targetKeyspace.compareAndSet(existingAttempt, attempt)) {
+                // Otherwise, start a new keyspace attempt replacing the existing one.
+                // This is ok, as any observers of the existing attempt will still complete.
+                if (existingAttempt != null && !existingAttempt.future.isDone()) {
+                    logger.warn("{} There is already an in progress attempt to set keyspace on this connection for keyspace '{}', beginning set keyspace for '{}'.  "
+                                    + "This can happen if you constantly change the keyspace at runtime.",
+                            this, existingAttempt.keyspace, keyspace);
+                    ksFuture.setException(new DriverException("Aborting attempt to set keyspace to " + keyspace + " since there is already an in flight attempt to set keyspace to " + existingAttempt.keyspace + "."));
```

Comment:
Should we create a specialized exception for this?   The other thing is perhaps we don't need the warning above since the code that handles the future failing will log an expected exception.